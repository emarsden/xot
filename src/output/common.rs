use crate::NameId;

/// Indentation: pretty-print XML or HTML.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct Indentation {
    /// A list of element names where indentation changes are suppressed.
    pub suppress: Vec<NameId>,
}

/// Parameters used when serializing tokens.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct TokenSerializeParameters {
    /// Elements that should have their text content be serialized as CDATA
    /// sections.
    pub cdata_section_elements: Vec<NameId>,

    /// Whether to escape the `>` character in text content. By default this is
    /// true, which means that `>` is escaped as `&gt;`. If you set this to true,
    /// `>` is not escaped, except for the special case of `]]>` outside of CDATA,
    /// which is mandated by the XML specification to always be escaped.
    pub unescaped_gt: bool,
}
