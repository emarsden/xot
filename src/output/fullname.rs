// when we render a element or attribute name, we need to render it with the
// right prefix information. we also want to keep track of prefix declarations.
// This data structure maintains this information and avoids having to wander
// the tree to get it.

use std::borrow::Cow;

use crate::{NameId, NamespaceId, PrefixId, Xot};

type Namespaces = Vec<(PrefixId, NamespaceId)>;

// this structure maintains everything we need to know about a name per element
#[derive(Debug)]
struct FullnameInfo {
    // all namespaces defined on this node
    defined_namespaces: Namespaces,
    // namespaces known on this node
    all_namespaces: Namespaces,
}

impl FullnameInfo {
    fn new(node_namespaces: Namespaces, current_fullname_info: &FullnameInfo) -> Self {
        let all_namespaces = current_fullname_info
            .all_namespaces
            .iter()
            .chain(node_namespaces.iter())
            .copied()
            .collect();
        Self {
            defined_namespaces: node_namespaces,
            all_namespaces,
        }
    }

    fn namespace_by_prefix(&self, prefix: PrefixId) -> Option<NamespaceId> {
        self.all_namespaces
            .iter()
            .rev()
            .find(|(p, _)| *p == prefix)
            .map(|(_, ns)| *ns)
    }

    fn prefixes_by_namespace(&self, namespace: NamespaceId) -> impl Iterator<Item = PrefixId> + '_ {
        self.all_namespaces
            .iter()
            .rev()
            .filter(move |(_, n)| *n == namespace)
            .map(|(p, _)| *p)
    }

    // look for the prefix. prefer the empty prefix, and if that isn't there, the
    // most recently defined prefix
    fn element_prefix_by_namespace(&self, xot: &Xot, namespace: NamespaceId) -> Option<PrefixId> {
        if self
            .prefixes_by_namespace(namespace)
            .any(|p| p == xot.empty_prefix())
        {
            Some(xot.empty_prefix())
        } else {
            self.prefixes_by_namespace(namespace).next()
        }
    }

    // look for the prefix, but only if it's not the empty prefix, as this is
    // for attributes which cannot be unprefixed and still in a namespace
    fn attribute_prefix_by_namespace(&self, xot: &Xot, namespace: NamespaceId) -> Option<PrefixId> {
        self.prefixes_by_namespace(namespace)
            .find(|&prefix| prefix != xot.empty_prefix())
    }
}

pub(crate) struct FullnameSerializer<'a> {
    xot: &'a Xot,
    stack: Vec<FullnameInfo>,
}

impl<'a> FullnameSerializer<'a> {
    pub(crate) fn new(xot: &'a Xot, defined_namespaces: Namespaces) -> Self {
        Self {
            xot,
            stack: vec![FullnameInfo {
                defined_namespaces: defined_namespaces.clone(),
                all_namespaces: defined_namespaces,
            }],
        }
    }

    pub(crate) fn push(&mut self, defined_namespaces: Namespaces) {
        // optimization; we don't need to recalculate anything if we
        // already have the same namespaces. the cost is that we need
        // to keep track of whether this node defined namespaces for pop as well.
        if defined_namespaces.is_empty() {
            return;
        }
        let current_fullname_info = self.stack.last().unwrap();
        self.stack
            .push(FullnameInfo::new(defined_namespaces, current_fullname_info));
    }

    // this is handy for the HTML rendering system, which insists some namespaces
    // should be in the empty prefix (xhtml, mathml, svg)
    pub(crate) fn add_empty_prefix(&mut self, namespace_id: NamespaceId) {
        let current_fullname_info = self.stack.last_mut().unwrap();
        let empty_entry = (self.xot.empty_prefix(), namespace_id);
        current_fullname_info.defined_namespaces.push(empty_entry);
        current_fullname_info.all_namespaces.push(empty_entry);
    }

    pub(crate) fn pop(&mut self, has_namespaces: bool) {
        if has_namespaces {
            self.stack.pop();
        }
    }

    fn top(&self) -> &FullnameInfo {
        self.stack.last().unwrap()
    }

    // get the fullname. if None, we cannot generate the fullname due to a missing
    // prefix
    pub(crate) fn fullname_element(&self, name_id: NameId) -> Option<Cow<'a, str>> {
        let namespace_id = self.xot.namespace_for_name(name_id);
        if namespace_id == self.xot.no_namespace_id {
            Some(Cow::Borrowed(self.xot.local_name_str(name_id)))
        } else {
            let prefix_id = self
                .top()
                .element_prefix_by_namespace(self.xot, namespace_id)?;
            if prefix_id == self.xot.empty_prefix() {
                Some(Cow::Borrowed(self.xot.local_name_str(name_id)))
            } else {
                let prefix = self.xot.prefix_str(prefix_id);
                Some(Cow::Owned(format!(
                    "{}:{}",
                    prefix,
                    self.xot.local_name_str(name_id)
                )))
            }
        }
    }

    pub(crate) fn fullname_attribute(&self, name_id: NameId) -> Option<Cow<'a, str>> {
        let namespace_id = self.xot.namespace_for_name(name_id);
        if namespace_id == self.xot.no_namespace_id {
            Some(Cow::Borrowed(self.xot.local_name_str(name_id)))
        } else {
            let prefix_id = self
                .top()
                .attribute_prefix_by_namespace(self.xot, namespace_id)?;
            let prefix = self.xot.prefix_str(prefix_id);
            Some(Cow::Owned(format!(
                "{}:{}",
                prefix,
                self.xot.local_name_str(name_id)
            )))
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_element_no_namespace() {
        let mut xot = Xot::new();

        let a = xot.add_name("a");
        let fullname_serializer = FullnameSerializer::new(&xot, vec![]);

        assert_eq!(
            fullname_serializer.fullname_element(a),
            Some(Cow::Borrowed("a"))
        );
    }

    #[test]
    fn test_attribute_no_namespace() {
        let mut xot = Xot::new();

        let a = xot.add_name("a");
        let fullname_serializer = FullnameSerializer::new(&xot, vec![]);

        assert_eq!(
            fullname_serializer.fullname_attribute(a),
            Some(Cow::Borrowed("a"))
        );
    }

    #[test]
    fn test_element_with_prefix() {
        let mut xot = Xot::new();

        let ns = xot.add_namespace("ns");
        let a = xot.add_name_ns("a", ns);
        let prefix = xot.add_prefix("p");
        let fullname_serializer = FullnameSerializer::new(&xot, vec![(prefix, ns)]);

        assert_eq!(
            fullname_serializer.fullname_element(a),
            Some(Cow::Owned("p:a".to_string()))
        );
    }

    #[test]
    fn test_attribute_with_prefix() {
        let mut xot = Xot::new();

        let ns = xot.add_namespace("ns");
        let a = xot.add_name_ns("a", ns);
        let prefix = xot.add_prefix("p");
        let fullname_serializer = FullnameSerializer::new(&xot, vec![(prefix, ns)]);

        assert_eq!(
            fullname_serializer.fullname_attribute(a),
            Some(Cow::Owned("p:a".to_string()))
        );
    }

    #[test]
    fn test_element_default_namespace() {
        let mut xot = Xot::new();

        let ns = xot.add_namespace("ns");
        let a = xot.add_name_ns("a", ns);
        let fullname_serializer = FullnameSerializer::new(&xot, vec![(xot.empty_prefix(), ns)]);

        assert_eq!(
            fullname_serializer.fullname_element(a),
            Some(Cow::Borrowed("a"))
        );
    }

    #[test]
    fn test_element_default_namespace_empty_prefix_preferred() {
        let mut xot = Xot::new();

        let ns = xot.add_namespace("ns");
        let a = xot.add_name_ns("a", ns);
        let p = xot.add_prefix("p");
        let fullname_serializer =
            FullnameSerializer::new(&xot, vec![(xot.empty_prefix(), ns), (p, ns)]);

        assert_eq!(
            fullname_serializer.fullname_element(a),
            Some(Cow::Borrowed("a"))
        );
    }

    #[test]
    fn test_element_most_recently_defined_prefix_preferred() {
        let mut xot = Xot::new();

        let ns = xot.add_namespace("ns");
        let a = xot.add_name_ns("a", ns);
        let p1 = xot.add_prefix("p1");
        let p2 = xot.add_prefix("p2");
        let fullname_serializer = FullnameSerializer::new(&xot, vec![(p1, ns), (p2, ns)]);

        assert_eq!(
            fullname_serializer.fullname_element(a),
            Some(Cow::Owned("p2:a".to_string()))
        );
    }

    #[test]
    fn test_element_missing_prefix() {
        let mut xot = Xot::new();

        let ns = xot.add_namespace("ns");
        let a = xot.add_name_ns("a", ns);
        let fullname_serializer = FullnameSerializer::new(&xot, vec![]);

        assert_eq!(fullname_serializer.fullname_element(a), None);
    }

    #[test]
    fn test_attribute_explicit_prefix_used_over_empty_prefix() {
        let mut xot = Xot::new();

        let ns = xot.add_namespace("ns");
        let a = xot.add_name_ns("a", ns);
        let p = xot.add_prefix("p");
        let fullname_serializer =
            FullnameSerializer::new(&xot, vec![(xot.empty_prefix(), ns), (p, ns)]);

        assert_eq!(
            fullname_serializer.fullname_attribute(a),
            Some(Cow::Owned("p:a".to_string()))
        );
    }

    #[test]
    fn test_attribute_explicit_prefix_used_over_empty_prefix2() {
        let mut xot = Xot::new();

        let ns = xot.add_namespace("ns");
        let a = xot.add_name_ns("a", ns);
        let p = xot.add_prefix("p");
        let fullname_serializer =
            FullnameSerializer::new(&xot, vec![(p, ns), (xot.empty_prefix(), ns)]);

        assert_eq!(
            fullname_serializer.fullname_attribute(a),
            Some(Cow::Owned("p:a".to_string()))
        );
    }

    #[test]
    fn test_attribute_without_explicit_prefix_default_not_found() {
        let mut xot = Xot::new();

        let ns = xot.add_namespace("ns");
        let a = xot.add_name_ns("a", ns);

        let fullname_serializer = FullnameSerializer::new(&xot, vec![(xot.empty_prefix(), ns)]);

        assert_eq!(fullname_serializer.fullname_attribute(a), None);
    }
}
