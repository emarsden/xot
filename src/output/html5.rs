//! html5 output method.
//!
//! The main entry point is [`Parameters`], which you can pass into various
//! serialization methods to control the output.
//!
//! You can use these parameters to control the HTML5 serialization, which you
//! can obtain using [`Xot::html5`].

use crate::NameId;

use super::Indentation;

/// Parameters for HTML generation.
///
/// This only supports HTML 5.
///
/// This follows the HTML5 serialization rules described in
/// https://www.w3.org/TR/xslt-xquery-serialization/
///
/// In summary:
///
/// - no-namespace and XHTML namespace is serialized as HTML tags
///
/// - Always use explicit close tags such as `</p>`, never use self-closing
///  tags such as `<p/>`.
///
/// - certain HTML tags are unclosed (void names such `br`, `meta`), i.e.
///   `<br>`
///
/// - mathml and svg are serialized with default namespace only, no prefix.
///
/// - certain HTML tags are rendered preformatted (`pre`, `script`)
///
/// - Indentation takes inline elements (phrasing content names such as `a`,
///   `span`, etc) into account. U
///
/// - Unknown HTML elements (that have no namespace or in the XHTML namespace) are
///   also treated as inline elements.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct Parameters {
    /// Pretty-print HTML, and a list of elements where this is suppressed.
    ///
    /// This recognizes inline (phrasing) elements.
    pub indentation: Option<Indentation>,
    /// Elements that should be serialized as CDATA sections.
    ///
    /// These should only be used for elements in non-XML content, like MathML
    /// or SVG.
    pub cdata_section_elements: Vec<NameId>,
    // TODO: character maps
}
