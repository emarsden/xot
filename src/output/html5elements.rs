use ahash::{HashSet, HashSetExt};

use crate::id::NameId;
use crate::xotdata::Xot;
use crate::NamespaceId;

// used to determine whether something is a HTML 5 element
const XHTML_NS: &str = "https://www.w3.org/1999/xhtml";
const MATHML_NS: &str = "http://www.w3.org/1998/Math/MathML";
const SVG_NS: &str = "http://www.w3.org/2000/svg";

#[derive(Debug)]
pub(crate) struct Html5Elements {
    xhtml_namespace_id: NamespaceId,
    mathml_namespace_id: NamespaceId,
    svg_namespace_id: NamespaceId,
    pub(crate) html5_names: HtmlNames,
    pub(crate) phrasing_content_names: HtmlNames,
    pub(crate) void_names: HtmlNames,
    pub(crate) formatted_names: HtmlNames,
    pub(crate) no_escape_names: HtmlNames,
}

#[derive(Debug)]
pub(crate) struct HtmlNames {
    xhtml_namespace_id: NamespaceId,
    ids: HashSet<NameId>,
    names: HashSet<String>,
}

impl HtmlNames {
    fn new(xot: &mut Xot, xhtml_namespace_id: NamespaceId, names: &[&str]) -> Self {
        let mut ids = HashSet::new();
        for name in names {
            // lowercase names, no namespace
            ids.insert(xot.add_name_ns(name, xot.no_namespace()));
            // uppercase names, no namespace
            ids.insert(xot.add_name_ns(&name.to_ascii_uppercase(), xot.no_namespace()));
            // lowercase names, XHTML namespace
            ids.insert(xot.add_name_ns(name, xhtml_namespace_id));
            // uppercase names, XHTML namespace
            ids.insert(xot.add_name_ns(&name.to_ascii_uppercase(), xhtml_namespace_id));
        }
        Self {
            xhtml_namespace_id,
            ids,
            names: names.iter().map(|name| name.to_string()).collect(),
        }
    }

    pub(crate) fn is_html_element(&self, xot: &Xot, name_id: NameId) -> bool {
        let namespace = xot.namespace_for_name(name_id);
        namespace == self.xhtml_namespace_id || namespace == xot.no_namespace()
    }

    pub(crate) fn matches(&self, xot: &Xot, name_id: NameId) -> bool {
        // if we match any of the known ids, we're done right away
        if self.ids.contains(&name_id) {
            return true;
        }
        // if this is not an HTML element, we know it isn't one of the HTML names
        if !self.is_html_element(xot, name_id) {
            return false;
        }
        // otherwise, we do a case-insensitive lookup of the local name
        let name = xot.local_name_str(name_id);
        // now lowercase the name and look it up
        let name = name.to_ascii_lowercase();
        self.names.contains(&name)
    }
}

impl Html5Elements {
    pub(crate) fn new(xot: &mut Xot) -> Self {
        let xhtml_namespace_id = xot.add_namespace(XHTML_NS);
        let mathml_namespace_id = xot.add_namespace(MATHML_NS);
        let svg_namespace_id = xot.add_namespace(SVG_NS);
        let html5_names = [
            "a",
            "abbr",
            "address",
            "area",
            "article",
            "aside",
            "audio",
            "b",
            "base",
            "bdi",
            "bdo",
            "blockquote",
            "body",
            "br",
            "button",
            "canvas",
            "caption",
            "cite",
            "code",
            "col",
            "colgroup",
            "command",
            "datalist",
            "dd",
            "del",
            "details",
            "dfn",
            "div",
            "dl",
            "dt",
            "em",
            "embed",
            "fieldset",
            "figcaption",
            "figure",
            "footer",
            "form",
            "h1",
            "h2",
            "h3",
            "h4",
            "h5",
            "h6",
            "head",
            "header",
            "hgroup",
            "hr",
            "html",
            "i",
            "iframe",
            "img",
            "input",
            "ins",
            "kbd",
            "keygen",
            "label",
            "legend",
            "li",
            "link",
            "map",
            "mark",
            "math",
            "menu",
            "meta",
            "meter",
            "nav",
            "noscript",
            "object",
            "ol",
            "optgroup",
            "option",
            "output",
            "p",
            "param",
            "pre",
            "progress",
            "q",
            "rp",
            "rt",
            "ruby",
            "s",
            "samp",
            "script",
            "section",
            "select",
            "small",
            "source",
            "span",
            "strong",
            "style",
            "sub",
            "summary",
            "sup",
            "table",
            "tbody",
            "td",
            "template",
            "textarea",
            "tfoot",
            "th",
            "thead",
            "time",
            "title",
            "tr",
            "track",
            "u",
            "ul",
            "var",
            "video",
            "wbr",
        ];
        let html5_names = HtmlNames::new(xot, xhtml_namespace_id, &html5_names);

        let void_names = [
            "area", "base", "br", "col", "embed", "hr", "img", "input", "keygen", "link", "meta",
            "param", "source", "track", "wbr",
            // extra elements not in the HTML5 spec but null in HTML 4
            "basefont", "frame", "isindex",
        ];

        let void_names = HtmlNames::new(xot, xhtml_namespace_id, &void_names);

        let phrasing_content_names = [
            "a", "abbr", "area", "audio", "b", "bdi", "bdo", "br", "button", "canvas", "cite",
            "code", "command", "datalist", "del", "dfn", "em", "embed", "i", "iframe", "img",
            "input", "ins", "kbd", "keygen", "label", "map", "mark", "math", "meter", "noscript",
            "object", "output", "progress", "q", "ruby", "s", "samp", "script", "select", "small",
            "span", "strong", "sub", "sup", "svg", "textarea", "time", "u", "var", "video", "wbr",
        ];
        let phrasing_content_names =
            HtmlNames::new(xot, xhtml_namespace_id, &phrasing_content_names);

        let formatted_names = ["pre", "script", "style", "title", "textarea"];
        let formatted_names = HtmlNames::new(xot, xhtml_namespace_id, &formatted_names);

        let no_escape_names = ["script", "style"];
        let no_escape_names = HtmlNames::new(xot, xhtml_namespace_id, &no_escape_names);
        Self {
            html5_names,
            xhtml_namespace_id,
            mathml_namespace_id,
            svg_namespace_id,
            void_names,
            phrasing_content_names,
            formatted_names,
            no_escape_names,
        }
    }

    pub(crate) fn is_html_element(&self, xot: &Xot, name_id: NameId) -> bool {
        let namespace = xot.namespace_for_name(name_id);
        self.is_html_namespace(xot, namespace)
    }

    pub(crate) fn must_be_serialized_unprefixed(&self, namespace: NamespaceId) -> bool {
        namespace == self.xhtml_namespace_id
            || namespace == self.mathml_namespace_id
            || namespace == self.svg_namespace_id
    }

    pub(crate) fn is_html_namespace(&self, xot: &Xot, namespace_id: NamespaceId) -> bool {
        namespace_id == self.xhtml_namespace_id || namespace_id == xot.no_namespace()
    }
}
