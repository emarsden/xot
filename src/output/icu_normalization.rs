/// Unicode normalization.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum NormalizationForm {
    /// Normalization Form C, using the rules specified in [Character Model for
    /// the World Wide Web 1.0:
    /// Normalization](https://www.w3.org/TR/xslt-xquery-serialization/#charmod-norm).
    Nfc,
    /// NFD specifies the serialized result will be in Normalization Form D, as
    /// specified in [UAX #15: Unicode Normalization
    /// Forms](https://www.w3.org/TR/xslt-xquery-serialization/#UNICODE-NORMALIZATION-FORM).
    Nfd,
    /// Normalization Form KC, as specified in [UAX #15: Unicode Normalization
    /// Forms](https://www.w3.org/TR/xslt-xquery-serialization/#UNICODE-NORMALIZATION-FORM).
    Nfkc,
    /// Normalization Form KD, as specified in [UAX #15: Unicode Normalization
    /// Forms](https://www.w3.org/TR/xslt-xquery-serialization/#UNICODE-NORMALIZATION-FORM).
    Nfkd,
    // TODO: fully normalized
}

// pub(crate) fn create_normalize(
//     provider: &impl icu_provider::AnyProvider,
//     normalization_form: Option<output::xml::NormalizationForm>,
// ) -> Result<Normalize, Error> {
//     if let Some(normalization_form) = normalization_form {
//         use crate::output::xml::NormalizationForm::*;
//         use icu::normalizer::{ComposingNormalizer, DecomposingNormalizer};

//         match normalization_form {
//             Nfc => {
//                 let normalizer = ComposingNormalizer::try_new_nfc_with_any_provider(provider)?;
//                 Ok(Box::new(move |content| {
//                     normalizer.normalize(content).into()
//                 }))
//             }
//             Nfd => {
//                 let normalizer = DecomposingNormalizer::try_new_nfd_with_any_provider(provider)?;
//                 Ok(Box::new(move |content| {
//                     normalizer.normalize(content).into()
//                 }))
//             }
//             Nfkc => {
//                 let normalizer = ComposingNormalizer::try_new_nfkc_with_any_provider(provider)?;
//                 Ok(Box::new(move |content| {
//                     normalizer.normalize(content).into()
//                 }))
//             }
//             Nfkd => {
//                 let normalizer = DecomposingNormalizer::try_new_nfkd_with_any_provider(provider)?;
//                 Ok(Box::new(move |content| {
//                     normalizer.normalize(content).into()
//                 }))
//             }
//         }
//     } else {
//         Ok(Box::new(|content| content.into()))
//     }
// }
