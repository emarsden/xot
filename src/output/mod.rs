//! Xot offers functionality to serialize XML data in different ways.
//!
//! This module lets you control serialization in various ways.
mod common;
mod fullname;
pub mod html5;
mod html5_serializer;
#[cfg(feature = "icu")]
mod icu_normalization;
mod normalizer;
mod pretty;
mod serializer;
pub mod xml;
mod xml_serializer;

pub use common::Indentation;
pub(crate) use fullname::FullnameSerializer;
pub(crate) use fullname::NamespaceDeclarations;
pub(crate) use html5_serializer::{Html5Elements, Html5Serializer};
pub use normalizer::{NoopNormalizer, Normalizer};
pub(crate) use pretty::Pretty;
pub use pretty::PrettyOutputToken;
pub(crate) use serializer::gen_outputs;
pub use serializer::{Output, OutputToken};
pub(crate) use xml_serializer::XmlSerializer;
