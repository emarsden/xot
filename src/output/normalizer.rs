use std::borrow::Cow;

#[cfg(doc)]
use crate::Xot;
#[cfg(feature = "icu")]
#[cfg(doc)]
use icu;

/// A normalizer.
///
/// Normalizes text and attribute values. You can pass a normalizer using
/// [`Xot::serialize_xml_string_with_normalizer`] and
/// [`Xot::serialize_xml_write_with_normalizer`].
///
#[cfg_attr(
    feature = "icu",
    doc = r##"
If you enable the `icu` feature, you can use the `icu` normalizers [`icu::normalizer::ComposingNormalizer`] and
[`icu::normalizer::DecomposingNormalizer`] as normalizers."##
)]
pub trait Normalizer {
    /// Given a piece of text, give back the normalized version.
    fn normalize<'a>(&self, content: Cow<'a, str>) -> Cow<'a, str>;

    // in theory we could use normalize_iter on ICU to avoid allocating a string, but
    // that gets too hairy for now
}

/// A normalizer that does nothing at all. This is used by default if
/// you don't specify a normalizer.
pub struct NoopNormalizer;

impl Normalizer for NoopNormalizer {
    #[inline]
    fn normalize<'a>(&self, content: Cow<'a, str>) -> Cow<'a, str> {
        content
    }
}
