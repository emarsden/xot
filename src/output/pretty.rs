use crate::output::Output;
use crate::xmlvalue::ValueType;
use crate::xotdata::{Node, Xot};
use crate::{NameId, Value};

/// Pretty output token
///
/// Like [`OutputToken`](`crate::output::OutputToken`) but with extra information for
/// pretty printing.
pub struct PrettyOutputToken {
    /// indentation level.
    pub indentation: usize,
    /// Whether the token is prefixed by a space character.
    pub space: bool,
    /// The token
    ///
    /// This is a fragment of XML like `"<p"`, `a="A"` or `"</p>"`.
    pub text: String,
    /// Whether the token is suffixed by a newline character.
    pub newline: bool,
}

// we need to track where we are in xml:space, so that we can know when to
// insert newlines and indentation
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum Space {
    Empty,
    Default,
    Preserve,
}

// The stack keeps track of where we are, and the xml space state. We are
// either in a mixed element (with text and subcontent) (in which case we don't
// do any indentation anymore, including for its descendants), or in an element
// without text, in which case we can potentially indent
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum StackEntry {
    Unmixed(Space),
    Mixed,
}

pub(crate) struct Pretty<'a, IsSuppressed, IsInline>
where
    IsSuppressed: Fn(NameId) -> bool,
    IsInline: Fn(NameId) -> bool,
{
    xot: &'a Xot,
    is_suppressed: IsSuppressed,
    is_inline: IsInline,
    // a list of element names where we don't do indentation for the immediate content
    // suppress: &'a [NameId],
    stack: Vec<StackEntry>,
}

impl<'a, IsSuppressed, IsInline> Pretty<'a, IsSuppressed, IsInline>
where
    IsSuppressed: Fn(NameId) -> bool,
    IsInline: Fn(NameId) -> bool,
{
    pub(crate) fn new(xot: &'a Xot, is_suppressed: IsSuppressed, is_inline: IsInline) -> Self {
        Pretty {
            xot,
            is_suppressed,
            is_inline,
            stack: Vec::new(),
        }
    }

    fn unmixed(&mut self, space: Space) {
        self.stack.push(StackEntry::Unmixed(space));
    }

    fn mixed(&mut self) {
        self.stack.push(StackEntry::Mixed);
    }

    fn in_mixed(&self) -> bool {
        self.stack.iter().any(|e| *e == StackEntry::Mixed)
    }

    fn in_space_preserve(&self) -> bool {
        for entry in self.stack.iter().rev() {
            match entry {
                StackEntry::Unmixed(Space::Preserve) => return true,
                StackEntry::Unmixed(Space::Default) => return false,
                StackEntry::Unmixed(Space::Empty) => (),
                StackEntry::Mixed => return false,
            }
        }
        false
    }

    fn pop(&mut self) {
        self.stack.pop();
    }

    fn get_indentation(&self) -> usize {
        if self.in_mixed() {
            return 0;
        }
        let mut count = 0;
        let mut in_preserve = false;
        for entry in self.stack.iter() {
            match entry {
                StackEntry::Unmixed(Space::Default) => {
                    in_preserve = false;
                    count += 1
                }
                StackEntry::Unmixed(Space::Preserve) => in_preserve = true,
                StackEntry::Unmixed(Space::Empty) => {
                    if !in_preserve {
                        count += 1
                    }
                }
                StackEntry::Mixed => (),
            }
        }
        count
    }

    fn get_newline(&self) -> bool {
        !self.in_mixed() && !self.in_space_preserve()
    }

    fn has_inline_child(&self, node: Node) -> bool {
        // a node has an inline child if it has a text node, or an element
        // defined as inline
        self.xot
            .children(node)
            .any(|child| match self.xot.value(child) {
                Value::Text(_) => true,
                Value::Element(element) => (self.is_inline)(element.name()),
                _ => false,
            })
    }

    fn element_space(&self, node: Node) -> Space {
        let attributes = self.xot.attributes(node);
        let space = attributes
            .get(self.xot.xml_space_name())
            .map(|s| s.as_str());
        match space {
            Some("preserve") => Space::Preserve,
            Some("default") => Space::Default,
            _ => Space::Empty,
        }
    }

    pub(crate) fn prettify(&mut self, node: Node, output_token: &Output) -> (usize, bool) {
        use Output::*;
        match output_token {
            StartTagOpen(_) => (self.get_indentation(), false),
            Comment(_) | ProcessingInstruction(..) => (self.get_indentation(), self.get_newline()),
            StartTagClose => {
                let newline = if self.xot.first_child(node).is_some() {
                    if !self.has_inline_child(node) {
                        let suppress = if let Some(element) = self.xot.element(node) {
                            (self.is_suppressed)(element.name())
                        } else {
                            false
                        };
                        // treat suppress as mixed content, as we don't want to indent
                        // anywhere inside
                        if suppress {
                            self.mixed();
                        } else {
                            let space = self.element_space(node);
                            self.unmixed(space);
                        }
                        self.get_newline()
                    } else {
                        self.mixed();
                        false
                    }
                } else {
                    false
                };
                (0, newline)
            }
            EndTag(_) => {
                let indentation = if self.xot.first_child(node).is_some() {
                    let no_indentation = self.in_mixed();
                    self.pop();
                    if !no_indentation {
                        self.get_indentation()
                    } else {
                        0
                    }
                } else {
                    0
                };
                (indentation, self.get_newline())
            }
            _ => (0, false),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;
    use rstest::rstest;

    use crate::output;

    #[rstest]
    fn pretty(
        #[values(
            ("elements", r#"<doc><a><b/></a></doc>"#, vec![]),
            ("more elements", r#"<doc><a><b/></a><a><b/><b/></a></doc>"#, vec![]),
            ("text", r#"<doc><a>text</a><a>text 2</a></doc>"#, vec![]),
            ("mixed", r#"<doc><p>Hello <em>world</em>!</p></doc>"#, vec![]),
            ("mixed, nested", r#"<doc><p>Hello <em><strong>world</strong></em>!</p></doc>"#, vec![]),
            ("mixed, multi", r#"<doc><p>Hello <em>world</em>!</p><p>Greetings, <strong>universe</strong>!</p></doc>"#, vec![]),
            // the embedded content could in principle be pretty printed, we don't. This is similar to xmllint --format behavior
            ("mixed, embedded", r#"<doc><p>Hello <nested><stuff>a</stuff><stuff>b</stuff></nested></p></doc>"#, vec![]),
            ("comment", r#"<doc><a><!--hello--><!--world--></a></doc>"#, vec![]),
            ("mixed, comment", r#"<doc><p>Hello <!--world-->!</p></doc>"#, vec![]),
            ("multi pi", r#"<doc><a><?pi?><?pi?></a></doc>"#, vec![]),
            ("preserve", r#"<doc xml:space="preserve"><p>Hello</p></doc>"#, vec![]),
            ("preserve_nested", r#"<doc xml:space="preserve">  <p><foo>  </foo></p></doc>"#, vec![]),
            ("preserve_back_to_default", r#"<doc xml:space="preserve"><p xml:space="default"><foo><bar/></foo></p></doc>"#, vec![]),
            ("not suppressed", r#"<doc><a><b/></a></doc>"#, vec![]),
            ("suppressed", r#"<doc><a><b/></a></doc>"#, vec!["a"]),
            ("suppressed nested", r#"<doc><a><b><c/></b></a></doc>"#, vec!["a"]),
        )]
        value: (&str, &str, Vec<&str>),
    ) {
        let (name, xml, suppress) = value;
        let mut xot = Xot::new();
        let suppress = suppress.iter().map(|s| xot.add_name(s)).collect();

        let document = xot.parse(xml).unwrap();
        let output_xml = xot
            .serialize_xml_string(
                output::xml::Parameters {
                    indentation: Some(output::Indentation { suppress }),
                    ..Default::default()
                },
                document,
            )
            .unwrap();
        assert_snapshot!(name, output_xml, xml);
    }
}
