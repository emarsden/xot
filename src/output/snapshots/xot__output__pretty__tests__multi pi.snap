---
source: src/output/pretty.rs
expression: "<doc><a><?pi?><?pi?></a></doc>"
---
<doc>
  <a>
    <?pi?>
    <?pi?>
  </a>
</doc>

